content/en/blog/istio-resiliency-patterns/index.md

```diff
@@ -0,0 +1,542 @@
+
+![Jaeger Trace Details](jaeger-tracedetails.png)
+
+Each trace in Jaeger has a set of spans that correspond to services that were invoked when a request has been made. To trace details, Jaeger also shows how long each span took - for example, in the above figure, the call to the greeter service took 17.05ms.
+
+You will also notice a bunch of other Istio related calls in the trace - such as a call to istio-policy and istio-mixer, etc.
+
+To make our services traceable, we need to propagate HTTP headers each time we're doing calls to another service. In the Hello web service, we are making calls to the greeter service, so we need to ensure the following headers are retrieved from the incoming request and added to the outgoing request:
+
+```sh
+x-request-id
+x-b3-traceid
+x-b3-spanid
+x-b3-parentspanid
+x-b3-sampled
+x-b3-flags
+x-ot-span-context
+```
+
+Here's the code snippet from the Hello web source code that deals with these headers:
+
+```go
+function getFowardHeaders(req) {
+    var headers = {};
+
+    for (var i = 0; i < 7; i++) {
+        var traceHeader = traceHeaders[i],
+            value = req.get(traceHeader);
+
+        if (value) {
+            headers[traceHeader] = value;
+        }
+    }
+    const useragent = req.get('user-agent');
+    headers['user-agent'] = useragent;
+    return headers;
+}
+
+const traceHeaders = [
+    'x-request-id',
+    'x-b3-traceid',
+    'x-b3-spanid',
+    'x-b3-parentspanid',
+    'x-b3-sampled',
+    'x-b3-flags',
+    'x-ot-span-context',
+];
+```
+
+The `getForwardHeaders` function is called whenever we are making requests to the downstream (greeter service) service: 
+
+```go
+const headers = getFowardHeaders(req);
+...
+axios.get(serviceUrl, { headers: headers, timeout: 5000 })
+...
+```
+
+## Istio Resiliency Features
+
+Istio gives you the ability to automatically set up timeouts, retries and circuit breakers, without touching or modifying your source code. This is a great feature you could use and ensure all your services are consistent. 
+
+### Request Timeouts
+
+Timeouts in Istio can be set in the virtual service resource using the `timeout` key. We are going to use a `v3` version of the greeter service - in this version of the service we have two endpoints: `/hello` and `/version`. The difference in this version is that the call to the `/version` endpoint takes an extra 2 seconds to return (we have a 2-second artificial delay in there). 
+
+To demonstrate the timeouts (and later retries), we will set up the timeout on the `v3` greeter service to `0.5` seconds - that way, the request will timeout because of the 2-second delay we have hard-coded. 
+
+Let's deploy the v3 version of the greeter service first:
+
+```sh
+kubectl create -f code/ch4/greeter-service-v3.yaml
```

Comment:
I think it might be better to put the contents of the YAML file here (also need to update the image names probably)